File: src/processor/transformer/code-block.transformer.ts
import MessageTransformer from './transformer';
import hljs from 'highlight.js';

class CodeBlockTransformer extends MessageTransformer {
  process() {
    return (content: string) => new Promise((resolve, reject) => {
      const regex = /```([a-z]*)\\n([\s\S]*?)```/gm;
      let m;

      console.log('CODE BLOCK', content);

      while ((m = regex.exec(content)) !== null) {
        console.log('M ', m);

        //let highlighted = rainbow.colorSync(m[2], m[1]);
        let highlightedValue;
        if (m[1] && m[1].length > 0) {
          highlightedValue = hljs.highlight(m[1], m[2]).value;
        } else {
          highlightedValue = hljs.highlightAuto(m[2]).value;
        }

        highlightedValue = `<pre>${highlightedValue}</pre>`;

        content = content.replace(m[0], highlightedValue);
        console.log('Replaced: ', content);
      }

      resolve(content);
    });

  }
}

export default CodeBlockTransformer;

File: backend/src/main/java/com/talkforgeai/talkforgeaiserver/transformers/CodeBlockTransformer.java
package com.talkforgeai.talkforgeaiserver.transformers;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class CodeBlockTransformer implements Transformer {

    @Override
    public String process(String content) {
        String regex = "```([a-z]*)\n([\\s\\S]*?)```";

        Pattern pattern = Pattern.compile(regex);
        Matcher matcher = pattern.matcher(content);

        while (matcher.find()) {
            String lang = matcher.group(1);
            String code = matcher.group(2);

            String processed = """
                    <pre><code class="language-%s">%s</code></pre>
                    """.formatted(lang, code);

            content = content.replace(matcher.group(), processed);
            System.out.println("Replaced: " + content);
        }

        return content;
    }
}
